gt/main.c

```diff
@@ -131,36 +130,85 @@ gt_parse_incoming_pkt(struct rte_mbuf *pkt, struct gt_packet_headers *info)
 		info->inner_ip_ver = ETHER_TYPE_IPv4;
 		info->l4_proto = inner_ipv4_hdr->next_proto_id;
 		info->l4_hdr = &inner_ipv4_hdr[1];
+
+		if (rte_ipv4_frag_pkt_is_fragmented(inner_ipv4_hdr)) {
+			info->frag = true;
+			info->l2_len = parsed_len;
+			info->l3_len = sizeof(*inner_ipv4_hdr);
+		}
 	} else if (likely(inner_ip_ver == 6)) {
-		/*
-	 	 * Make sure that the packet has space for
-		 * at least 4 bytes for the l4 header.
-	 	 */
-		if (pkt->data_len < parsed_len + sizeof(struct ipv6_hdr) + 4)
+		if (pkt->data_len < parsed_len + sizeof(struct ipv6_hdr))
 			return -1;
 
 		inner_ipv6_hdr = (struct ipv6_hdr *)info->inner_l3_hdr;
 		info->inner_ip_ver = ETHER_TYPE_IPv6;
 		info->l4_proto = inner_ipv6_hdr->proto;
 		info->l4_hdr = &inner_ipv6_hdr[1];
+
+		/*
+		 * XXX The function rte_ipv6_frag_get_ipv6_fragment_header()
+		 * only looks at the extension header that's right after
+		 * the fixed IPv6 header, and doesn't follow the whole chain
+		 * of extension headers.
+		 */
+		info->frag_hdr =
+			rte_ipv6_frag_get_ipv6_fragment_header(inner_ipv6_hdr);
+
+		if (info->frag_hdr != NULL) {
+			info->frag = true;
+			info->l2_len = parsed_len;
+			info->l3_len = sizeof(*inner_ipv6_hdr) +
+				sizeof(*info->frag_hdr);
+		}
 	} else
 		return -1;
 
 	return 0;
 }
 
+static struct rte_mbuf *
+gt_reassemble_incoming_pkt(struct rte_mbuf *pkt,
+	uint64_t tms, struct gt_packet_headers *info,
+	struct rte_ip_frag_death_row *death_row, struct gt_instance *instance)
+{
+	/* Prepare mbuf: setup l2_len/l3_len. */
+	pkt->l2_len = info->l2_len;
+	pkt->l3_len = info->l3_len;
+
+	if (info->inner_ip_ver == ETHER_TYPE_IPv4) {
+		/* Process this IPv4 fragment. */
+		return rte_ipv4_frag_reassemble_packet(
+			instance->frag_tbl, death_row,
+			pkt, tms, info->inner_l3_hdr);
+	}
+
+	if (likely(info->inner_ip_ver == ETHER_TYPE_IPv6)) {
+		/* Process this IPv6 fragment. */
+		return rte_ipv6_frag_reassemble_packet(
+			instance->frag_tbl, death_row,
+			pkt, tms, info->inner_l3_hdr, info->frag_hdr);
+	}
+
+	rte_panic("Unexpected condition: gt at lcore %u reassembles a packet with unknown IP version %hu\n",
+		rte_lcore_id(), info->inner_ip_ver);
+
+	return NULL;
+}
+
 static int
 lookup_policy_decision(struct gt_packet_headers *pkt_info,
 	struct ggu_policy *policy, struct gt_instance *instance)
 {
 	policy->flow.proto = pkt_info->inner_ip_ver;
 	if (pkt_info->inner_ip_ver == ETHER_TYPE_IPv4) {
-		struct ipv4_hdr *ip4_hdr = (struct ipv4_hdr *)pkt_info->inner_l3_hdr;
+		struct ipv4_hdr *ip4_hdr =
+			(struct ipv4_hdr *)pkt_info->inner_l3_hdr;
 
 		policy->flow.f.v4.src = ip4_hdr->src_addr;
 		policy->flow.f.v4.dst = ip4_hdr->dst_addr;
 	} else if (likely(pkt_info->inner_ip_ver == ETHER_TYPE_IPv6)) {
-		struct ipv6_hdr *ip6_hdr = (struct ipv6_hdr *)pkt_info->inner_l3_hdr;
+		struct ipv6_hdr *ip6_hdr =
+			(struct ipv6_hdr *)pkt_info->inner_l3_hdr;
```

Comment:
Were these (or the IPv4 case above) lines changed? If not, they can fit on one line (and reduce the size of the pull request).

Reply:
These lines were not changed, but they exceed the 80 characters per line limit.